src/server/pfs/server/driver_new_storage.go

```diff
@@ -85,45 +83,67 @@ func (d *driver) finishCommitNewStorageLayer(txnCtx *txnenv.TransactionContext,
 	return d.writeFinishedCommit(txnCtx.Stm, commit, commitInfo)
 }
 
-// (bryce) should have a context going through here.
-func (d *driver) putFilesNewStorageLayer(pachClient *client.APIClient, server *putFileServer) error {
+func (d *driver) putFilesNewStorageLayer(pachClient *client.APIClient, server *putFileServer) (retErr error) {
 	// (bryce) how to handle no commit started?
-	// (bryce) this will not work for large files with the current implementation.
-	for req, err := server.Recv(); err == nil; req, err = server.Recv() {
-		if err := d.fs.WriteHeader(&tar.Header{Name: req.File.Path}); err != nil {
-			return err
-		}
-		if _, err := io.Copy(d.fs, bytes.NewReader(req.Value)); err != nil {
-			return err
+	// (bryce) a failed put (crash/error) should result with any serialized sub file sets getting
+	// cleaned up.
+	// (bryce) piping is not ideal, but easy for getting integration started.
+	// we will want to eventually just read raw bytes from the grpc stream.
```

Comment:
I think we have most of the code required to not use a pipe in the `putFileReader` type, you should be able to just take that and maybe modify it a little to get it to work here.

Reply:
This has been refactored.